diff --git a/src/google/protobuf/compiler/command_line_interface.cc b/src/google/protobuf/compiler/command_line_interface.cc
index a5b7402..3f104d3 100644
--- a/src/google/protobuf/compiler/command_line_interface.cc
+++ b/src/google/protobuf/compiler/command_line_interface.cc
@@ -278,9 +278,17 @@ void AddDefaultProtoPaths(
     return;
   }
   // Check if the upper level directory has an "include" subdirectory.
-  pos = path.find_last_of("/\\");
-  if (pos == std::string::npos || pos == 0) {
-    return;
+  // In vcpkg we need to skip 2 levels. The directory structure is
+  //    .../bin/tools/protoc
+  //    .../include/
+  // as opposed to
+  //    .../bin/protoc
+  //    .../include/
+  for (int i = 0; i != 2; ++i) {
+    pos = path.find_last_of("/\\");
+    if (pos == std::string::npos || pos == 0) {
+      return;
+    }
   }
   path = path.substr(0, pos);
   include_path = absl::StrCat(path, "/include");
