diff --git a/upb/json/decode.c b/upb/json/decode.c
index 51a0937..dd8fe00 100644
--- a/upb/json/decode.c
+++ b/upb/json/decode.c
@@ -1176,7 +1176,7 @@ static void jsondec_struct(jsondec* d, upb_Message* msg,
 static void jsondec_wellknownvalue(jsondec* d, upb_Message* msg,
                                    const upb_MessageDef* m) {
   upb_MessageValue val;
-  const upb_FieldDef* f;
+  const upb_FieldDef* f = 0;
   upb_Message* submsg;
 
   switch (jsondec_peek(d)) {
